internal/tools/orchestrator/hepa/providers/domain/domain.service.go

```diff
@@ -117,9 +122,63 @@ func (s *domainService) ChangeInnerIngress(ctx context.Context, req *pb.ChangeIn
 	}
 	return
 }
+
+func (s *domainService) preCheckPermission(userID string, services *bundle.GetRuntimeServicesResponseData) error {
+	if services == nil || services.Extra == nil {
+		return errors.New("can't get runtime extra info")
+	}
+
+	var (
+		appId        = services.Extra.ApplicationId
+		resourceType = s.getPermissionResource(services)
+	)
+
+	if userID == "" {
+		return apierrors.ErrGetUser
+	}
+	if access, err := s.bdl.CheckPermission(&apistructs.PermissionCheckRequest{
+		UserID:   userID,
+		Scope:    apistructs.AppScope,
+		ScopeID:  appId,
+		Action:   apistructs.GetAction,
+		Resource: resourceType,
+	}); err != nil || !access.Access {
+		return apierrors.ErrGetProject.AccessDenied()
+	}
+	return nil
+}
+
+func (s *domainService) getPermissionResource(services *bundle.GetRuntimeServicesResponseData) string {
+	return "runtime-" + strings.ToLower(services.Extra.Workspace)
```

Comment:
It's more appropriate to use `render` or `compose` instead of `get` for function parameters that have already been validated, such as `workspace`.